internal/eventstream/api_responses.go

```diff
@@ -0,0 +1,287 @@
+}
+
+// Downtime represents the Icinga 2 API Downtime object.
+//
+// NOTE:
+//   - An empty Service field indicates a host service.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#objecttype-downtime
+type Downtime struct {
+	Host    string `json:"host_name"`
+	Service string `json:"service_name"`
+	Author  string `json:"author"`
+	Comment string `json:"comment"`
+}
+
+// HostServiceRuntimeAttributes are common attributes of both Host and Service objects.
+//
+// When catching up potentially missed changes, the following fields seems to be the relevant ones which, fortunately,
+// are identical for Icinga 2 Host and Service objects.
+//
+// NOTE:
+//   - Name is either the Host or the Service name.
+//   - Host is empty for Host objects; Host contains the Service's Host object name for Services.
+//   - State might be 0 = UP, 1 = DOWN for hosts and 0 = OK, 1 = WARNING, 2 = CRITICAL, 3 = UNKNOWN for services.
+//   - StateType might be 0 = SOFT, 1 = HARD.
+//   - Acknowledgement type is 0 = NONE, 1 = NORMAL, 2 = STICKY.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#host
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#service
+type HostServiceRuntimeAttributes struct {
+	Name                      string      `json:"name"`
+	Host                      string      `json:"host_name,omitempty"`
+	Groups                    []string    `json:"groups"`
+	State                     int         `json:"state"`
+	StateType                 int         `json:"state_type"`
+	LastCheckResult           CheckResult `json:"last_check_result"`
+	LastStateChange           Icinga2Time `json:"last_state_change"`
+	DowntimeDepth             int         `json:"downtime_depth"`
+	Acknowledgement           int         `json:"acknowledgement"`
+	AcknowledgementLastChange Icinga2Time `json:"acknowledgement_last_change"`
+}
+
+// ObjectQueriesResult represents the Icinga 2 API Object Queries Result wrapper object.
+//
+// The Attrs field will be populated with a pointer to the representing type after JSON unmarshalling.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#object-queries-result
+type ObjectQueriesResult struct {
+	Name  string `json:"name"`
+	Type  string `json:"type"`
+	Attrs any    `json:"attrs"`
+}
+
+func (objQueriesRes *ObjectQueriesResult) UnmarshalJSON(bytes []byte) error {
+	var responseAttrs json.RawMessage
+	err := json.Unmarshal(bytes, &struct {
+		Name  *string          `json:"name"`
+		Type  *string          `json:"type"`
+		Attrs *json.RawMessage `json:"attrs"`
+	}{&objQueriesRes.Name, &objQueriesRes.Type, &responseAttrs})
+	if err != nil {
+		return err
+	}
+
+	switch objQueriesRes.Type {
+	case "Comment":
+		objQueriesRes.Attrs = new(Comment)
+	case "Downtime":
+		objQueriesRes.Attrs = new(Downtime)
+	case "Host", "Service":
+		objQueriesRes.Attrs = new(HostServiceRuntimeAttributes)
+	default:
+		return fmt.Errorf("unsupported type %q", objQueriesRes.Type)
+	}
```

Comment:
For the `/v1/objects` API queries you know (in contrast to the event stream) what type you're expecting in the response. So it should be possible two avoid the two pass unmarshal here as well as some type assertions elsewhere. The most invasive change this requires is that functions like `queryObjectsApi()` have to know the type, maybe a case where generics could come in handy.